### PR TITLE
internal/ethapi: add celo_dynamic_tx support to TransactionArgs

### PR DESCRIPTION
**Pull request by YoduYodu**
### Description

Added celo_dynamic_tx support to TransactionArgs.

### Other changes

Refactor code in `setDefaults()`:
    - tidy up if statement
    - remove unreachable code

### Tested

Currently the e2e transfer tests fail bizarrely, in order to be able to test Espresso activation, a list of test cases will be listed here and ran manually through mycelo + celocli/curl as a transient approach.

The test net has an Espresso HF block activated in the middle of two set of pre and post Espresso targeted transactions.
All transactions are sent in the form of below with fields modified,
```
curl -X POST -H "Content-Type: application/json" \
--data "{\"jsonrpc\":\"2.0\", \"method\":\"eth_sendTransaction\", \"params\":[{
    \"from\": \"${validator1}\",
    \"to\":\"${validator2}\",
    \"gas\": \"0x15f90\",
    \"maxFeePerGas\": \"0x77359400\",
    \"value\": \"0x1FEB3DD067660000\"
    }], \"id\": 267}" \
http://localhost:8545
```
And using `celocli account:balance ADDRESS` to check the balance.

#### Test cases

Pre-Espresso:

- Should reject 0x1 ("transaction type not supported") ✅ 
- Should reject 0x2 (error "maxFeePerGas or maxPriorityFeePerGas specified but Espresso is not active yet") ✅ 
- Should reject 0x7c (error "maxFeePerGas or maxPriorityFeePerGas specified but Espresso is not active yet") ✅ 
- Should execute 0x0 (ETH-compatible) transfer CELO, fee CELO (consumed value + gas, value received) ✅ 
- Should execute 0x0 (ETH-compatible) transfer CELO, fee cUSD (consumed CELO value, consumed cUSD gas, CELO value received) ✅ 
- Should execute 0x0 (ETH-compatible) transfer cUSD, fee CELO (consumed cUSD value , consumed CELO gas, cUSD value received) ✅ 
- Should execute 0x0 (ETH-compatible) transfer cUSD, fee cUSD (consumed cUSD value + gas, cUSD value received) ✅ 
- Should execute 0x0 (ETH-incompatible) transfer CELO, fee CELO (consumed value + gas, value received, gatewayFee received) ✅ 
- Should execute 0x0 (ETH-incompatible) transfer CELO, fee cUSD (consumed CELO value, consumed cUSD gas, CELO value received, gatewayFee received) ✅ 
- Should execute 0x0 (ETH-incompatible) transfer cUSD, fee CELO (consumed cUSD value , consumed CELO gas, cUSD value received, gatewayFee received) ✅ 
- Should execute 0x0 (ETH-incompatible) transfer cUSD, fee cUSD (consumed cUSD value + gas, cUSD value received, gatewayFee received) ✅ 

Post-Espresso:

- Should execute 0x1 ✅ 
- Should execute 0x2, transfer CELO, fee CELO (consumed value + gas, value received) ✅ 
- Should execute 0x7c transfer CELO, fee CELO (consumed value + gas, value received, gateway fee received) ✅ 
- Should execute 0x0 (ETH-compatible) transfer CELO, fee CELO (consumed value + gas, value received) ✅ 
- Should execute 0x0 (ETH-incompatible) transfer CELO, fee CELO (consumed value + gas, value received) ✅ 
- Should execute 0x7c transfer CELO, fee CELO (consumed value + gas, value received) ✅ 
- Should execute 0x7c transfer CELO, fee cUSD (consumed CELO value, consumed cUSD gas, CELO value received) ✅ 
- Should execute 0x7c transfer cUSD, fee CELO (consumed cUSD value , consumed CELO gas, cUSD value received) ✅ 
- Should execute 0x7c transfer cUSD, fee cUSD (consumed cUSD value + gas, cUSD value received) ✅ 
- Should execute 0x7c transfer CELO, fee CELO (consumed value + gas, value received, gatewayFee received) ✅ 
- Should execute 0x7c transfer CELO, fee cUSD (consumed CELO value, consumed cUSD gas, CELO value received, gatewayFee received) ✅ 
- Should execute 0x7c transfer cUSD, fee CELO (consumed cUSD value , consumed CELO gas, cUSD value received, gatewayFee received) ✅ 
- Should execute 0x7c transfer cUSD, fee cUSD (consumed cUSD value + gas, cUSD value received, gatewayFee received) ✅ 

### Backwards compatibility

Part of Espresso HF.

### CODE REVIEW

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 7979f2f91d88b07233c71d0d83ea23721d5c7d7c
<details><summary>coverage:  44.7% of statements across all listed packages</summary><br>
<pre>
coverage:  54.1% of statements in consensus/istanbul
coverage:  45.0% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  43.3% of statements in consensus/istanbul/backend/internal/db
coverage:  24.1% of statements in consensus/istanbul/backend/internal/enodes
coverage:  22.4% of statements in consensus/istanbul/backend/internal/replica
coverage:  64.0% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
</pre>
CommentID: 49558055ea
</details>

File: internal/ethapi/transaction_args.go [hunk 1]
```diff
@@ -122,16 +124,6 @@ func (args *TransactionArgs) setDefaults(ctx context.Context, b Backend) error {
 				if err != nil {
 					return err
 				}
-				if b.ChainConfig().IsEspresso(head.Number) {
```
Comment (trianglesphere) [inline internal/ethapi/transaction_args.go:125]: Above this price should probably be the suggested gas price, not to the suggested gas tip cap (thought to be low relative to the basefee)

File: internal/ethapi/transaction_args.go [hunk 2]
```diff
@@ -276,16 +268,33 @@ func (args *TransactionArgs) toTransaction() *types.Transaction {
...
-			To:         args.To,
-			ChainID:    (*big.Int)(args.ChainID),
-			Nonce:      uint64(*args.Nonce),
-			Gas:        uint64(*args.Gas),
-			GasFeeCap:  (*big.Int)(args.MaxFeePerGas),
-			GasTipCap:  (*big.Int)(args.MaxPriorityFeePerGas),
-			Value:      (*big.Int)(args.Value),
-			Data:       args.data(),
-			AccessList: al,
+		if args.FeeCurrency != nil || args.GatewayFeeRecipient != nil {
```
Comment (trianglesphere) [inline internal/ethapi/transaction_args.go:271]: At minimum, if any celo specific field is present, it should use the celo dynamic fee tx.

**Review by trianglesphere — COMMENTED** [review]
Some of my comments are more at the code around this than the specific changes you've made.

It might also add the fee currency in the gas estimation. Using a non-native FC does incur an additional cost, but I'm not fully sure how it's accounted for.

File: internal/ethapi/transaction_args.go [hunk 3]
```diff
@@ -91,27 +91,29 @@ func (args *TransactionArgs) setDefaults(ctx context.Context, b Backend) error {
 	// need to consult the chain for defaults. It's definitely a London tx.
 	if args.MaxPriorityFeePerGas == nil || args.MaxFeePerGas == nil {
 		// In this clause, user left some fields unspecified.
-		if b.ChainConfig().IsEspresso(head.Number) && (args.GasPrice == nil || args.GasPrice.ToInt().Cmp(big.NewInt(0)) == 0) {
-			if args.MaxPriorityFeePerGas == nil {
-				tip, err := b.SuggestGasTipCap(ctx, args.FeeCurrency)
-				if err != nil {
-					return err
+		if b.ChainConfig().IsEspresso(head.Number) {
+			if args.GasPrice == nil || args.GasPrice.ToInt().Cmp(big.NewInt(0)) == 0 {
```
[the remark this answers is not in the document]
Comment (mcortesi) [reply internal/ethapi/transaction_args.go:95]: After checking the logic, this seems to be fine

**Review by mcortesi — APPROVED** [review]
LGTM

File: internal/ethapi/transaction_args.go (hunk 2 above)
Comment (YoduYodu) [reply internal/ethapi/transaction_args.go:271]: https://github.com/celo-org/celo-blockchain/blob/7979f2f91d88b07233c71d0d83ea23721d5c7d7c/internal/ethapi/transaction_args.go#L180
if `args.GatewayFeeRecipient != nil`, then `args.GatewayFee` must have been set

Comment (trianglesphere) [reply internal/ethapi/transaction_args.go:271]: Hmm. That's true locally, but I don't think that invariant elsewhere. This is something that works, but I'm not sure it's worth the effort to be clever vs just adding another check.

**Comment by trianglesphere** [comment]
It needs to pass gas price & other info to the gas estimation
```
@@ -165,6 +167,9 @@ func (args *TransactionArgs) setDefaults(ctx context.Context, b Backend) error {
                        GasPrice:             args.GasPrice,
                        MaxFeePerGas:         args.MaxFeePerGas,
                        MaxPriorityFeePerGas: args.MaxPriorityFeePerGas,
+                       FeeCurrency:          args.FeeCurrency,
+                       GatewayFee:           args.GatewayFee,
+                       GatewayFeeRecipient:  args.GatewayFeeRecipient,
```

**Pull request merged**